stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt

```diff
@@ -83,53 +85,56 @@ internal class ChannelControllerImpl(
-    private val _endOfNewerMessages = MutableLiveData(false)
-    private val _endOfOlderMessages = MutableLiveData(false)
-    private val _loading = MutableLiveData(false)
-    private val _hidden = MutableLiveData(false)
-    private val _muted = MutableLiveData(false)
-    private val _watchers = MutableLiveData<Map<String, User>>(mapOf())
-    private val _members = MutableLiveData<Map<String, Member>>()
-    private val _loadingOlderMessages = MutableLiveData(false)
-    private val _loadingNewerMessages = MutableLiveData(false)
-    private val _channelData = MutableLiveData<ChannelData>()
-    private val _oldMessages = MutableLiveData<Map<String, Message>>()
+    private val _messages = MutableStateFlow<Map<String, Message>?>(null)
+    private val _watcherCount = MutableStateFlow<Int>(0)
+    private val _typing = MutableStateFlow<Map<String, ChatEvent>>(emptyMap())
+    private val _reads = MutableStateFlow<Map<String, ChannelUserRead>>(emptyMap())
+    private val _read = MutableStateFlow<ChannelUserRead?>(null)
+    private val _endOfNewerMessages = MutableStateFlow(false)
+    private val _endOfOlderMessages = MutableStateFlow(false)
+    private val _loading = MutableStateFlow(false)
+    private val _hidden = MutableStateFlow(false)
+    private val _muted = MutableStateFlow(false)
+    private val _watchers = MutableStateFlow<Map<String, User>>(emptyMap())
+    private val _members = MutableStateFlow<Map<String, Member>>(emptyMap())
+    private val _loadingOlderMessages = MutableStateFlow(false)
+    private val _loadingNewerMessages = MutableStateFlow(false)
+    private val _channelData = MutableStateFlow<ChannelData?>(null)
+    private val _oldMessages = MutableStateFlow<Map<String, Message>?>(emptyMap())
     internal var hideMessagesBefore: Date? = null
-    val unfilteredMessages: LiveData<List<Message>> =
-        Transformations.map(_messages) { it.values.toList() }
+    val unfilteredMessages: LiveData<List<Message>?> = _messages.map {
+        if (it == null) { null } else { it.values.toList() }
+    }.asLiveData()
 
     /** a list of messages sorted by message.createdAt */
     override val messages: LiveData<List<Message>> = messagesTransformation(_messages)
 
     override val oldMessages: LiveData<List<Message>> = messagesTransformation(_oldMessages)
 
-    private fun messagesTransformation(messages: MutableLiveData<Map<String, Message>>): LiveData<List<Message>> {
-        return Transformations.map(messages) { messageMap ->
-            messageMap.values
-                .asSequence()
-                .filter { it.parentId == null || it.showInChannel }
-                .filter { hideMessagesBefore == null || it.wasCreatedAfter(hideMessagesBefore) }
-                .sortedBy { it.createdAt ?: it.createdLocallyAt }
-                .toList()
-                .map { it.copy() }
+    private fun messagesTransformation(messages: MutableStateFlow<Map<String, Message>?>): LiveData<List<Message>> {
```

Comment:
From what I understand so far, LiveData always allows null as a value. So if you tell LiveData the type is String it's actually String?. See this blogpost for instance: https://proandroiddev.com/improving-livedata-nullability-in-kotlin-45751a2bafb7

Reply:
Solved this and removed a bunch of nullability in general in https://github.com/GetStream/stream-chat-android/pull/824/commits/e00f2c91a64e68decfcf8c5372cffc6063a1cf8c

Reply:
So how do you now distinguish between messages being:

- unknown
- empty
- a list of messages

You need some sort of state for that first one (previously represented as null)

Reply:
I tried looking up the original usages of these `LiveData`s, but I didn't find anywhere in the codebase where we handle a `null` value being emitted from them. Looks to me like they get to `MessageListItemLiveData` and we assume them to be non-null there.

Reply:
LiveData always emits null, so if we don't handle it that can be a problem

Reply:
It can technically emit it, but only if we assign that value to it. Looks to me like we're okay with nulls in the current state of the PR, the changes to `Flow` should make things type safe.

Reply:
Ok it needs a bit of integration/testing with the UI side of things to make sure that we handle the "messages unknown" state well.